gtsam/geometry/Cal3Fisheye.h

```diff
@@ -0,0 +1,211 @@
+/* ----------------------------------------------------------------------------
+
+ * GTSAM Copyright 2010, Georgia Tech Research Corporation,
+ * Atlanta, Georgia 30332-0415
+ * All Rights Reserved
+ * Authors: Frank Dellaert, et al. (see THANKS for the full author list)
+
+ * See LICENSE for the license information
+
+ * -------------------------------------------------------------------------- */
+
+/**
+ * @file Cal3Fisheye.h
+ * @brief Calibration of a fisheye camera
+ * @date Apr 8, 2020
+ * @author ghaggin
+ */
+
+#pragma once
+
+#include <gtsam/geometry/Point2.h>
+
+namespace gtsam {
+
+/**
+ * @brief Calibration of a fisheye camera
+ * @addtogroup geometry
+ * \nosubgrouping
+ *
+ * Uses same distortionmodel as OpenCV, with
+ * https://docs.opencv.org/master/db/d58/group__calib3d__fisheye.html
+ * 3D point in camera frame
+ *   p = (x, y, z)
+ * Intrinsic coordinates:
+ *   [x_i;y_i] = [x/z; y/z]
+ * Distorted coordinates:
+ *   r^2 = (x_i)^2 + (y_i)^2
+ *   th = atan(r)
+ *   th_d = th(1 + k1*th^2 + k2*th^4 + k3*th^6 + k4*th^8)
+ *   [x_d; y_d] = (th_d / r)*[x_i; y_i]
+ * Pixel coordinates:
+ *   K = [fx s u0; 0 fy v0 ;0 0 1]
+ *   [u; v; 1] = K*[x_d; y_d; 1]
+ */
+class GTSAM_EXPORT Cal3Fisheye {
+ protected:
+  double fx_, fy_, s_, u0_, v0_;  // focal length, skew and principal point
+  double k1_, k2_, k3_, k4_;      // fisheye distortion coefficients
+
+ public:
+  enum { dimension = 9 };
+  typedef boost::shared_ptr<Cal3Fisheye>
+      shared_ptr;  ///< shared pointer to fisheye calibration object
+
+  /// @name Standard Constructors
+  /// @{
+
+  /// Default Constructor with only unit focal length
+  Cal3Fisheye()
+      : fx_(1), fy_(1), s_(0), u0_(0), v0_(0), k1_(0), k2_(0), k3_(0), k4_(0) {}
+
+  Cal3Fisheye(const double fx, const double fy, const double s, const double u0,
+              const double v0, const double k1, const double k2,
```

Comment:
I'm not sure about your appetite on working more on this PR, but I'm wondering whether we really need scalar access, or whether we could just work with a Vector4 all over the place, and get rid of the k1...k4 access etc below.